File: packages/python/src/typeflux/project/annotations.py
"""Insight acknowledgement annotations (#733 §1, #577): an in-repo, PR-reviewed
record of which console insights have been acknowledged/suppressed, and why.

The maintainer decision (#577, 2026-07-17) is that ack/suppression state lives in
the repo — git stays the single source of truth, PR-reviewed like everything else;
there is NO console-side or control-plane-side mutable state. This module is the
read side of that decision: it parses ``.typeflux/annotations.yaml`` (beside the
project manifest) into a normalized, contract-shaped projection the read tier
serves, and it feeds the validation surface so a malformed file is an authoring
error rather than a silent drop.

Shape of the file (a MAPPING, like every other Typeflux YAML — extensible, and its
``annotations`` field mirrors the served projection envelope exactly)::

    annotations:
      - insight_id_pattern: "policy.drift.*"     # exact id or glob
        reason: "tracked upstream; not actionable until the provider ships the fix"
        tracked_in: "https://github.com/acme/infra/issues/412"   # optional issue URL
        expires: 2026-12-31                                       # optional ISO date

* ``insight_id_pattern`` matches the console's stable insight ids (exact or glob);
  it is required and must be a non-empty, trimmed string.
* ``reason`` is required (audit honesty: an ack without a reason is not an ack).
* ``tracked_in`` is optional and, when present, must be an http(s) URL — the SHAPE
  is validated (scheme + host), never fetched (no network in the read tier).
* ``expires`` is an optional ISO date. An expired entry is STILL SERVED with its
  expiry (rendering — the "stale ack" insight — is the console's job, slice 3); the
  projection only carries what the console needs to derive staleness.

Parsing is **fail-closed** (never partial/silent). Three outcomes:

* **absent file** → an EMPTY projection, NOT an error (the common case — most
  projects acknowledge nothing).
* **valid file** → the parsed annotations, in file order.
* **malformed/unparseable file** → an EMPTY projection AND a recorded parse error.
  The projection stays empty (never a half-parsed subset); the error is surfaced as
  a project VALIDATION ISSUE by :func:`typeflux.project.loader.validate_project`
  (an authoring error on the read/validation surface — NOT an enforcement verdict,
  keeping the explicit enforcement code-set untouched, #723).

:func:`read_project_annotations` is the single parse (both the projection and the
validation issue derive from it, so they can never disagree); :func:`load_project_annotations`
is the projection-only convenience.
"""

from __future__ import annotations

from datetime import date
from pathlib import Path
from typing import NamedTuple
from urllib.parse import urlparse

from pydantic import BaseModel, ConfigDict, field_validator

from typeflux.project.spec import TypefluxProjectSpec
from typeflux.yaml.loader import strict_safe_load

#: The annotations file's location relative to the project manifest: ``.typeflux/``
#: beside the manifest, matching the design (#733). A dot-directory keeps the ack
#: ledger out of the workflow-file namespace while staying in-repo and PR-reviewed.
ANNOTATIONS_DIRNAME = ".typeflux"
ANNOTATIONS_FILENAME = "annotations.yaml"


class InsightAnnotation(BaseModel):
    """One acknowledged/suppressed insight entry (#733). ``insight_id_pattern`` and
    ``reason`` are required; ``tracked_in`` (shape-validated issue URL) and ``expires``
    (ISO date) are optional. Frozen + ``extra=forbid``: an unknown key is an authoring
    error, surfaced as a validation issue rather than silently ignored."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    insight_id_pattern: str
    reason: str
    tracked_in: str | None = None
    expires: date | None = None

    @field_validator("insight_id_pattern", "reason")
    @classmethod
    def _non_empty_trimmed(cls, value: str) -> str:
        if not value or value.strip() != value:
            raise ValueError("must be a non-empty, trimmed string")
        return value

    @field_validator("tracked_in")
    @classmethod
    def _validate_tracked_in(cls, value: str | None) -> str | None:
        # SHAPE only — an http(s) URL with a host. The read tier never fetches it
        # (no network); the console renders it as a link (slice 3).
        if value is None:
            return value
        if not value or value.strip() != value:
            raise ValueError("tracked_in must be a non-empty, trimmed URL")
        try:
            parsed = urlparse(value)
        except ValueError as exc:
            raise ValueError(f"tracked_in is not a valid URL: {exc}") from exc
        if parsed.scheme not in ("http", "https") or not parsed.netloc:
            raise ValueError("tracked_in must be an http(s) URL citing the tracking issue")
        return value


class ProjectAnnotations(BaseModel):
    """The served annotations projection envelope (#733): every acknowledgement entry,
    in file order. Empty for a project with no annotations file (the common case) or a
    malformed one (fail-closed — the malformed file is a validation issue instead)."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    annotations: tuple[InsightAnnotation, ...] = ()


class _AnnotationsFile(BaseModel):
    """The on-disk file shape: a mapping with an ``annotations`` list. ``extra=forbid``
    so a typo'd top-level key (``annotation:``) fails loudly instead of yielding an
    empty-but-valid projection that silently drops every entry."""

    model_config = ConfigDict(extra="forbid")

    annotations: tuple[InsightAnnotation, ...] = ()


class AnnotationsReadResult(NamedTuple):
    """The single parse both consumers derive from. ``annotations`` is always a valid
    (possibly empty) projection — never a partial parse; ``error`` is ``None`` on
    success or absence, and a human-readable parse message when the file exists but is
    malformed (the validation-issue source); ``path`` is the resolved file location
    (for the issue's ``path`` field)."""

    annotations: ProjectAnnotations
    error: str | None
    path: Path


def annotations_path(project: TypefluxProjectSpec) -> Path:
    """The project's annotations file location: ``<manifest_dir>/.typeflux/annotations.yaml``."""
    return project.project_dir / ANNOTATIONS_DIRNAME / ANNOTATIONS_FILENAME


def read_project_annotations(project: TypefluxProjectSpec) -> AnnotationsReadResult:
    """Parse the project's annotations file, fail-closed (#733).

    Absent file → empty projection, no error (the common case). A YAML/schema failure
    → empty projection AND a parse-error message (the projection is never a partial
    subset). An empty/comments-only file (``None``) → empty projection, no error.
    """
    path = annotations_path(project)
    if not path.exists():
        return AnnotationsReadResult(ProjectAnnotations(), None, path)
    try:
        raw = strict_safe_load(path.read_text(encoding="utf-8"))
    except Exception as exc:  # noqa: BLE001 - any load failure is an authoring error, reported.
        return AnnotationsReadResult(
            ProjectAnnotations(), f"annotations file is not valid YAML: {exc}", path
        )
    if raw is None:
        # An empty or comments-only file is an empty ledger, not a malformed one.
        return AnnotationsReadResult(ProjectAnnotations(), None, path)
    try:
        parsed = _AnnotationsFile.model_validate(raw)
    except Exception as exc:  # noqa: BLE001 - pydantic ValidationError (+ any coercion error).
        return AnnotationsReadResult(
            ProjectAnnotations(), f"annotations file does not match the schema: {exc}", path
        )
    return AnnotationsReadResult(ProjectAnnotations(annotations=parsed.annotations), None, path)


def load_project_annotations(project: TypefluxProjectSpec) -> ProjectAnnotations:
    """The annotations projection the read tier serves (#733): the parsed entries, or an
    empty projection when the file is absent or malformed (fail-closed). The malformed
    case additionally surfaces as a validation issue via ``validate_project``."""
    return read_project_annotations(project).annotations


__all__ = [
    "ANNOTATIONS_DIRNAME",
    "ANNOTATIONS_FILENAME",
    "AnnotationsReadResult",
    "InsightAnnotation",
    "ProjectAnnotations",
    "annotations_path",
    "load_project_annotations",
    "read_project_annotations",
]

File: packages/typescript/temporal-controlplane/src/annotations.ts
/**
 * Insight acknowledgement annotations (#733 §1; Python `project/annotations.py`): the read side of
 * the in-repo insight-ack ledger. Parses `.typeflux/annotations.yaml` (beside the project manifest)
 * into the contract-shaped projection the read tier serves, and reports a parse error so the
 * validation surface can raise an AUTHORING issue for a malformed file.
 *
 * The maintainer decision (#577) keeps ack/suppression state IN THE REPO — PR-reviewed, git as the
 * single source of truth, no console-side or control-plane-side mutable state. So this is a
 * pure-YAML, project-level read with NO seam and NO resolution dependency: always servable, even for
 * a project this server cannot resolve.
 *
 * Parsing is FAIL-CLOSED, byte-for-byte with the Python edition (behavioral parity):
 *  - absent file        → an EMPTY projection, NOT an error (the common case).
 *  - empty/comments-only → an EMPTY projection, NOT an error.
 *  - valid file          → the parsed entries, in file order.
 *  - malformed/invalid   → an EMPTY projection AND a parse-error message (never a partial parse);
 *                          the error surfaces as a project validation issue, not here.
 *
 * An expired entry is SERVED with its expiry (the "stale ack" rendering is the console's job, slice
 * 3) — expiry never mutates or hides an entry.
 */

import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";

import { parseDocument } from "yaml";
import { z } from "zod";

import type { components } from "./http/contract.js";

/** The served projection envelope (Python `ProjectAnnotations`), aliased off the generated contract
 * so a schema regen that changes the shape breaks this compile instead of drifting the edition. */
export type ApiProjectAnnotations = components["schemas"]["ProjectAnnotations"];
/** One acknowledgement entry (Python `InsightAnnotation`). */
export type ApiInsightAnnotation = components["schemas"]["InsightAnnotation"];

/** The annotations file location relative to the manifest: `<manifestDir>/.typeflux/annotations.yaml`
 * (Python `annotations_path`). */
export const ANNOTATIONS_DIRNAME = ".typeflux";
export const ANNOTATIONS_FILENAME = "annotations.yaml";

const nonEmptyTrimmed = (value: string): boolean => value.length > 0 && value.trim() === value;

/** SHAPE-only URL validation (Python `_validate_tracked_in`): an http(s) URL with a host. The read
 * tier NEVER fetches it. */
function isHttpUrl(value: string): boolean {
  if (!nonEmptyTrimmed(value)) return false;
  let parsed: URL;
  try {
    parsed = new URL(value);
  } catch {
    return false;
  }
  return (parsed.protocol === "http:" || parsed.protocol === "https:") && parsed.host.length > 0;
}

/** `expires` is an ISO calendar date (`YYYY-MM-DD`) — matching Python's `datetime.date` acceptance.
 * The YAML core schema leaves the scalar a string (no timestamp tag), so validate the string shape
 * and that it names a real date. */
function isIsoDate(value: string): boolean {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return false;
  const [year, month, day] = value.split("-").map((part) => Number(part));
  const date = new Date(Date.UTC(year!, month! - 1, day!));
  return (
    date.getUTCFullYear() === year && date.getUTCMonth() === month! - 1 && date.getUTCDate() === day
  );
}

const insightAnnotationSchema = z
  .object({
    insight_id_pattern: z
      .string()
      .refine(nonEmptyTrimmed, "insight_id_pattern must be a non-empty, trimmed string"),
    reason: z.string().refine(nonEmptyTrimmed, "reason must be a non-empty, trimmed string"),
    // `.nullish()` + null→undefined: YAML authors write `tracked_in:` / explicit nulls, and
    // Python's pydantic optionals treat explicit null as absent — the editions must accept the
    // same files (codex: a null here invalidated the WHOLE file on this edition only).
    tracked_in: z
      .string()
      .refine(isHttpUrl, "tracked_in must be an http(s) URL citing the tracking issue")
      .nullish()
      .transform((value) => value ?? undefined),
    expires: z
      .string()
      .refine(isIsoDate, "expires must be an ISO date (YYYY-MM-DD)")
      .nullish()
      .transform((value) => value ?? undefined),
  })
  .strict();

const annotationsFileSchema = z
  .object({
    annotations: z.array(insightAnnotationSchema).default([]),
  })
  .strict();

/** The single parse both consumers (the projection + the validation issue) derive from, so they can
 * never disagree. `error` is `undefined` on success/absence, a human-readable message when the file
 * exists but is malformed. */
export interface AnnotationsReadResult {
  annotations: ApiProjectAnnotations;
  error?: string;
  path: string;
}

/** The annotations file path for a manifest (Python `annotations_path`). */
export function annotationsPath(manifestPath: string): string {
  return join(dirname(manifestPath), ANNOTATIONS_DIRNAME, ANNOTATIONS_FILENAME);
}

/** Parse the project's annotations file, fail-closed (Python `read_project_annotations`). */
export function readProjectAnnotations(manifestPath: string): AnnotationsReadResult {
  const path = annotationsPath(manifestPath);
  if (!existsSync(path)) {
    return { annotations: { annotations: [] }, path };
  }
  let raw: unknown;
  try {
    // `merge: true` expands YAML `<<` merge keys, matching Python's strict_safe_load and the
    // temporal-yaml loader — without it a merge key reaches the strict schema and invalidates
    // the whole file on this edition only (codex).
    const doc = parseDocument(readFileSync(path, "utf-8"), { uniqueKeys: true, merge: true });
    const fatal = [...doc.errors, ...doc.warnings.filter((w) => w.code === "DUPLICATE_KEY")];
    if (fatal.length > 0) {
      return {
        annotations: { annotations: [] },
        error: `annotations file is not valid YAML: ${fatal[0]?.message}`,
        path,
      };
    }
    raw = doc.toJS();
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    return { annotations: { annotations: [] }, error: `annotations file is not valid YAML: ${message}`, path };
  }
  if (raw === null || raw === undefined) {
    // Empty or comments-only file: an empty ledger, not a malformed one.
    return { annotations: { annotations: [] }, path };
  }
  const parsed = annotationsFileSchema.safeParse(raw);
  if (!parsed.success) {
    return {
      annotations: { annotations: [] },
      error: `annotations file does not match the schema: ${parsed.error.message}`,
      path,
    };
  }
  // Rebuild each entry so absent optionals are OMITTED (never `undefined`-valued keys) — parity with
  // Python's `exclude_none` serialization, so the wire body matches byte-for-byte.
  const annotations: ApiInsightAnnotation[] = parsed.data.annotations.map((entry) => {
    const out: ApiInsightAnnotation = {
      insight_id_pattern: entry.insight_id_pattern,
      reason: entry.reason,
    };
    if (entry.tracked_in !== undefined) out.tracked_in = entry.tracked_in;
    if (entry.expires !== undefined) out.expires = entry.expires;
    return out;
  });
  return { annotations: { annotations }, path };
}

/** The annotations projection the read tier serves (Python `load_project_annotations`): the parsed
 * entries, or an empty projection when the file is absent or malformed (fail-closed). */
export function loadProjectAnnotations(manifestPath: string): ApiProjectAnnotations {
  return readProjectAnnotations(manifestPath).annotations;
}
